test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ContainmentTest/Server/ContainmentTestController.cs

```diff
@@ -0,0 +1,516 @@
+
+                if (paymentInstrument == null)
+                {
+                    return NotFound();
+                }
+
+                myPaymentInstruments.Remove(paymentInstrument);
+            }
+
+            return NoContent();
+        }
+
+        [EnableQuery]
+        [HttpGet("odata/Accounts({key})/Default.GetDefaultPI()")]
+        public IActionResult GetDefaultPI([FromRoute] int key)
+        {
+            var account = _dataSource.Accounts.SingleOrDefault(a => a.AccountID == key);
+
+            if (account == null)
+            {
+                return NotFound();
+            }
+
+            if (account.MyPaymentInstruments != null && account.MyPaymentInstruments.Count > 0)
+            {
+                PaymentInstrument pi = account.MyPaymentInstruments[0];
+
+                return Ok(pi);
+            }
+
+            return null;
+        }
+
+        [HttpPost("odata/Accounts({key})/Default.RefreshDefaultPI")]
+        public IActionResult RefreshDefaultPI([FromRoute] int key, ODataActionParameters parameters)
+        {
+            var account = _dataSource.Accounts.SingleOrDefault(a => a.AccountID == key);
+
+            if (account == null)
+            {
+                return NotFound();
+            }
+
+            if (account.MyPaymentInstruments != null && account.MyPaymentInstruments.Count > 0)
+            {
+                PaymentInstrument pi = account.MyPaymentInstruments[0];
+                pi.CreatedDate = (DateTimeOffset)parameters["newDate"];
+
+                return Ok(pi);
+            }
+
+            return null;
+        }
+
+        [EnableQuery]
+        [HttpPatch("odata/Accounts({key})/MyPaymentInstruments({paymentInstrumentId})")]
+        public IActionResult UpdateAccountMyPaymentInstrument([FromRoute] int key, [FromRoute] int paymentInstrumentId, [FromBody] PaymentInstrument updatedPaymentInstrument)
+        {
+            var account = _dataSource.Accounts.SingleOrDefault(a => a.AccountID == key);
+
+            if (account == null)
+            {
+                return NotFound();
+            }
+
+            var existingPaymentInstrument = account.MyPaymentInstruments?.SingleOrDefault(pi => pi.PaymentInstrumentID == paymentInstrumentId);
+
+            if (existingPaymentInstrument == null)
+            {
+                account.MyPaymentInstruments ??= new List<PaymentInstrument>();
+                account.MyPaymentInstruments.Add(updatedPaymentInstrument);
+            }
+            else
+            {
+                existingPaymentInstrument.FriendlyName = updatedPaymentInstrument.FriendlyName;
+                existingPaymentInstrument.CreatedDate = updatedPaymentInstrument.CreatedDate;
+            }
+
+            var result = account.MyPaymentInstruments.SingleOrDefault(pi => pi.PaymentInstrumentID == paymentInstrumentId);
+
+            return Updated(result);
+        }
+
+
+        [HttpPost("odata/containmenttests/Default.ResetDefaultDataSource")]

```

Comment:
Is this a non-OData endpoint or how does the `/containmenttests` segment get resolved? Or does it work just okay for a non-bound `ResetDefaultDataSource` action? Can you confirm that it'd work if it the endpoint URI was invoked

Reply:
Yes this works. 

Doing this from the test class works ok: 

```c#

 private void ResetDefaultDataSource()
        {
            var actionUri = new Uri(_baseUri + "containmenttests/Default.ResetDefaultDataSource", UriKind.Absolute);
            _context.Execute(actionUri, "POST");
        }

```